sdks/java/harness/src/main/java/org/apache/beam/fn/harness/debug/OutputSampler.java

```diff
@@ -129,20 +158,43 @@ public List<BeamFnApi.SampledElement> samples() throws IOException {
       resampleIndex = 0;
     }
 
+    // An element can live in both the main samples and exception buffer. Use a small look up table
+    // to deduplicate samples.
+    HashSet<Long> seen = new HashSet<>();
     ByteStringOutputStream stream = new ByteStringOutputStream();
     for (int i = 0; i < bufferToSend.size(); i++) {
       int index = (sampleIndex + i) % bufferToSend.size();
+      ElementSample<T> sample = bufferToSend.get(index);
+      seen.add(sample.id);
+
+      if (valueCoder != null) {
+        this.valueCoder.encode(sample.sample.getValue(), stream, Coder.Context.NESTED);
+      } else if (windowedValueCoder != null) {
+        this.windowedValueCoder.encode(sample.sample, stream, Coder.Context.NESTED);
+      }
```

Comment:
Would it make sense to add tests for the following cases? Feel free to just resolve this comment if you don't agree as you are closer to the code than I am. It looked like your new tests covered the valueCoder != null instantiating OutputSampler with a `VarIntCoder`, already. 

| valueCoder | windowedValueCoder |
| --- | --- |
| null | null |
| not null | null |
| null | not null |
| not null | not null |

Reply:
I don't think so or at least it would be prohibitively weird to test. These coders are controlled by the constructor and are a one-of. And for those cases there are already tests.